src/Services/DatabaseBackup/MysqlCustomDatabaseBackup.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+/*
+ * This file is part of the Liip/FunctionalTestBundle
+ *
+ * (c) Lukas Kahwe Smith <smith@pooteeweet.org>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Liip\FunctionalTestBundle\Services\DatabaseBackup;
+
+use Doctrine\Common\DataFixtures\Executor\AbstractExecutor;
+use Doctrine\ORM\EntityManager;
+
+/**
+ * @author Aleksey Tupichenkov <alekseytupichenkov@gmail.com>
+ *
+ * It's class created just for example that how to create database backup/restore service
+ */
+class MysqlCustomDatabaseBackup extends AbstractDatabaseBackup
+{
+    protected static $referenceData;
+
+    protected static $sql;
+
+    protected static $metadata;
+
+    public function getBackupName(): string
+    {
+        return $this->container->getParameter('kernel.cache_dir').'/test_mysql_'.md5(serialize($this->metadatas).serialize($this->classNames)).'.sql';
+    }
+
+    public function getReferenceBackupName(): string
+    {
+        return $this->getBackupName().'.ser';
+    }
+
+    protected function getBackup(): string
+    {
+        if (empty(self::$sql)) {
+            self::$sql = file_get_contents($this->getBackupName());
+        }
+
+        return self::$sql;
+    }
+
+    protected function getReferenceBackup(): string
+    {
+        if (empty(self::$referenceData)) {
+            self::$referenceData = file_get_contents($this->getReferenceBackupName());
+        }
+
+        return self::$referenceData;
+    }
+
+    public function isBackupActual(): bool
+    {
+        return
+            file_exists(file_exists($this->getBackupName())) &&
+            file_exists(file_exists($this->getReferenceBackupName())) &&
+            $this->isBackupUpToDate($this->getBackupName());
+    }
+
+    public function backup(AbstractExecutor $executor): void
+    {
+        $params = $this->connection->getParams();
+        if (isset($params['master'])) {
+            $params = $params['master'];
+        }
+
+        $dbName = isset($params['dbname']) ? $params['dbname'] : '';
+        $dbHost = isset($params['host']) ? $params['host'] : '';
+        $dbPort = isset($params['port']) ? $params['port'] : '';
+        $dbUser = isset($params['user']) ? $params['user'] : '';
+        $dbPass = isset($params['password']) ? $params['password'] : '';
+
+        $executor->getReferenceRepository()->save($this->getBackupName());
+        /** @var EntityManager $em */
+        $em = $executor->getReferenceRepository()->getManager();
+        self::$metadata = $em->getMetadataFactory()->getLoadedMetadata();
+
+        exec("mysqldump -h $dbHost -u $dbUser -p$dbPass --no-create-db $dbName > {$this->getBackupName()}");
```

Comment:
https://github.com/liip/LiipFunctionalTestBundle/pull/398#issuecomment-366554219

Reply:
@alexislefebvre @lsmith77 
About https://github.com/liip/LiipFunctionalTestBundle/pull/398#issuecomment-366554219, it's not suit, this method save only reference data, but I need save database data (sql with inserts data per each table)

Reply:
Indeed ..

Generally I think if we have to use a CLI command, then its not really a good fit for this Bundle, since it will not be applicable in too many cases.

This library here might help however https://github.com/portphp/doctrine

Reply:
Hmm... I added mysql class just for example
I can use `portphp/doctrine` or something else, it will be interesting for me, but I have fears about performance and additional vendor...

Currently I rename service `mysql_custom` -> `mysql_example` and add attention block from readme

As a proposal, mysql (and other) can be implemented by a separate repo in github, what do you think about it? @alexislefebvre @lsmith77

Reply:
@alekseytupichenkov 

> About #398 (comment), it's not suit, this method save only reference data, but I need save database data (sql with inserts data per each table)

The method works on the current version of this bundle. I'm sorry but I don't understand why you don't use it in this PR. Could you please explain why you chose to perform a full export instead of using only the reference data?

Reply:
@alexislefebvre I save reference too
https://github.com/liip/LiipFunctionalTestBundle/pull/398/files#diff-f2e540017ed6e7479f8d4a4b1e17a480R82

But it's not enough, I need backup both, DB and reference.
In case of SQLite, to copy DB enough copy just file, in case of MySql we can't just copy files, need make backup, for make backup I use `mysqldump` because Doctrine can't do it

Reply:
@lsmith77 about https://github.com/portphp/doctrine

This lib really cool, but still not suitable, this lib saves only entities, but I need dump full db.
For example, after loading fixtures I can have some additional data, by reason to Doctrine lifecycle methods or DB triggers

I found some PHP libs to make backup like https://github.com/backup-manager/backup-manager or https://github.com/spatie/db-dumper
But they all use `mysqldump` too

Reply:
Thanks for your explanation, I was confused by the fact that SQLite and MySQL need different ways to perform backups. I understand now. Thanks.